### DISCUSSION TITLE
Sharing cache between different requests

### DISCUSSION THREAD

**Discussion by Wassap124** [TanStack/query#5567] [General] [answered]
Hey!
Asking here cause I'm surely not the first to have tackled this.

Say I wanna get myself `posts` using `GET /posts`, and later, i wanna do `GET /posts/1`,
There's no real reason to fetch that post, since i'd already got it in cache, it's just a cache key management thing, right?

Also, say I create a new post using `POST /posts`. 
I want to update `GET /posts` cache in advance, and add that new post to it, since I know it should be added there

So I wanted to ask, before i try to reinvent the wheel, if anyone had tackled this before, and if you have a good solution for such a use case.

**Comment by samhirtarif** [comment]
You should be able to use [getQueryData](https://tanstack.com/query/v4/docs/react/reference/QueryClient#queryclientgetquerydata) for the first case, just get the cached data for the `/posts` key and fetch the relevant item?

And [setQueryData](https://tanstack.com/query/v4/docs/react/reference/QueryClient#queryclientsetquerydata) for the second case. After creating a new post you can update the data for the `GET /posts` by using the `setQueryData` on with the key you used for `GET /posts` and the new data

**Reply by Wassap124** [reply]
Thanks for replying!

I don't think I want to use `getQueryData`.
The most common use case I see in my mind is 
- I fetch all `posts`.
- I show a list of posts, meaning I already have them
- I press on a post to enter it's "page"
- the UI would try to fetch THAT particular post
- BUT - I already have that
- So i want, that in case someone does a fetch of a single post, to be able to assert if i already "have" it or not

So my general idea (and correct me if you think it's abusing `RQ`), is after I fetch all posts, to use `setQueriesData` for all queries that fetch `posts/1`, `/posts/2` etc.

**Reply by Wassap124** [reply]
I want that when a person would try to fetch a single post, and in the case it was included in the "bigger" fetch, it would be seamless for them

**Reply by TkDodo** [reply, marked as answer]
Have a read here: https://tkdodo.eu/blog/seeding-the-query-cache#seeding-details-from-lists

**Reply by samhirtarif** [reply]
^ The link that TkDodo shared explains what I was saying in much more detail and with an example 👍🏻

**Reply by Wassap124** [reply]
Thanks @TkDodo!
I guess a lot of people encounter the seeding use-case, huh?
Wonder how there's no auxiliary package no-one's made yet for structuring cache seeding in RQ

**Reply by Wassap124** [reply]
Im certain im not the first, nor the only person that encounters this issue - where I want to sync up different caches.
don't you think that a lot (if not most) users would require this ability?

Also - should i really do this to every different entity type I have? that could reach a good few dozens of different entity types.

again - It sounds to me like a lot of repeated code that a lot of ReactQuery users would appreciate

**Reply by TkDodo** [reply]
> I guess a lot of people encounter the seeding use-case, huh?

not really, no. Usually, the `list` queries don't have all the information that you need for the detail queries - it's a subset. Seeding only works if the list endpoints returns the exact same structure as the detail endpoint. At that point - why would you need a detail endpoint at all?

```
const useTodos = (select) => useQuery({
  queryKey: ['todos'],
  queryFn: fetchTodos,
  select
})

const useTodo = (id) => useTodos(todos => todos.find(t => t.id === id))
```

with this, you'll get:

```
useTodos() // gives you a list
useTodo(1) // gives you the first todo
```

**Reply by Wassap124** [reply]
> not really, no. Usually, the list queries don't have all the information that you need for the detail queries - it's a subset.

You really think so? A lot of people use `REST` which basically always includes full entities.

> why would you need a detail endpoint at all

That's a fair point. in my app, you can also access separate todos via the route, so if a user opens the my app on `/todos/1`, the thing that makes sense is to request for that resource - unless I already have it.

**Reply by Wassap124** [reply]
Also relevant for mutations.
say im adding a new todo - add it to the existing cached array instead of `refetch`ing

**Reply by Wassap124** [reply]
also just to make sure - sorry if I sound mad / condescending / demanding 😁
tone of speak is hard to pass via text.
 
im honestly asking everything 😅 
it just seems to me like a really common use-case, 
both at my current work and on previous projects i'd worked on this pattern seem to repeat

**Reply by Wassap124** [reply]
also a lot of reasons to fetch a single entity instead of all of them, when you know you only need the one.
instead of overfetching, decrease backend / db loads

**Reply by TkDodo** [reply]
>  in my app, you can also access separate todos via the route, so if a user opens the my app on /todos/1, the thing that makes sense is to request for that resource - unless I already have it.

yeah that also makes sense.

> say im adding a new todo - add it to the existing cached array inside of refetching

here, refetching actually makes more sense imo. You'd want the todo to show up in _all_ lists, for example, if you have filters or sorting, that would be part of the QueryKey, and writing that new todo to the right point in all those list cache entries is pretty cumbersome. Invalidating all todo lists solves the problem, and only fetches what you're currently using.

> both at my current work and on previous projects i'd worked on this pattern seem to repeat

I do see room in 3rd party plugins to support that use-case, but react-query itself is pretty agnostic to how cache entries are linked to each other, and it will likely stay that way.

**Reply by Wassap124** [reply]
> but react-query itself is pretty agnostic to how cache entries are linked to each other, and it will likely stay that way

Totally legit, thought it was worth asking as im bound to not be the only one 

If anyone encountering this thread and knows any 3rd party plugins that support this use-case, i'd be happy to hear about them.

Thanks for the attention and the quick replies @TkDodo
